tempesta_fw/t/fuzzer.c

```diff
@@ -365,54 +435,63 @@ __add_field(TfwFuzzContext *ctx, char **p, char *end, int t, int n)
 		}
 
 		if (r == FUZZ_INVALID)
-			TFW_DBG("generate ivalid random field for header %d\n",
+			TFW_DBG("generate invalid random field for header %d\n",
 				t);
 
 		return r;
 	}
 }
 
 static unsigned int
-add_field(TfwFuzzContext *ctx, char **p, char *end, int t)
+add_field(TfwFuzzContext *ctx, int type, char **p, char *end, int t)
 {
-	return __add_field(ctx, p, end, t, ctx->i[t]);
+	return __add_field(ctx, type, p, end, t, ctx->i[t]);
 }
 
 static unsigned int
-__add_header(TfwFuzzContext *ctx, char **p, char *end, int t, int n)
+__add_header(TfwFuzzContext *ctx, int type, char **p, char *end, int t, int n)
 {
 	unsigned int v = 0, i;
-	char *key;
 
 	BUG_ON(t < 0);
 	BUG_ON(t >= TRANSFER_ENCODING_NUM);
 
-	key = keys[t];
+	BUG_ON(!fld_data[t].key);
 
-	BUG_ON(!key);
-
-	add_string(p, end, key);
-	v |= add_field(ctx, p, end, SPACES);
-	v |= add_field(ctx, p, end, t);
+	add_string(p, end, fld_data[t].key);
+	v |= add_field(ctx, type, p, end, SPACES);
+	v |= add_field(ctx, type, p, end, t);
 	for (i = 0; i < n; ++i) {
 		addch(p, end, ',');
-		v |= add_field(ctx, p, end, SPACES);
-		v |= __add_field(ctx, p, end, t, (i * 256) %
+		v |= add_field(ctx, type, p, end, SPACES);
+		v |= __add_field(ctx, type, p, end, t, (i * 256) %
 			(gen_vector[t].size + gen_vector[t].over));
 	}
 
 	add_string(p, end, "\r\n");
 
 	if (v & FUZZ_INVALID)
-		TFW_DBG("generate ivalid header %d\n", t);
+		TFW_DBG("generate invalid header %d\n", t);
+
+	ctx->hdr_flags |= 1 << t;
 
 	return v;
 }
 
 static unsigned int
-add_header(TfwFuzzContext *ctx, char **p, char *end, int t)
+__add_header_rand(TfwFuzzContext *ctx, int type,
+		  char **p, char *end, int t, int n)
+{
+	unsigned int rand;
+
+	get_random_bytes(&rand, sizeof(rand));
```

Comment:
The fuzzer must be deterministic. I.e. if you run it and face an issue you should be able to reproduce the issue. If you use random gnerator, then it's hard to reproduce the fuzzer results. This is not ultimate fuzzing, but rather it's handy tool to efficiently generate various HTTP message variants which are hard to code all by hands. Also y we don't want to exhaust system entropy. Probably continuously incrementing static variable would be good. I reckon compiler optimizes `% 2` to `& 0x1`, but probably it's better to avoid div operation.